crates/sui-move-build/src/linters/redundant_custom_call.rs

```diff
@@ -0,0 +1,265 @@
+    #[default]
+    Other,
+}
+
+pub struct ExecutionContext {
+    diags: Diagnostics,
+}
+
+#[derive(Clone, Debug)]
+pub struct State {
+    locals: BTreeMap<Var, LocalState<Value>>,
+}
+
+//**************************************************************************************************
+// impls
+//**************************************************************************************************
+
+impl SimpleAbsIntConstructor for RedundantCustomVerifier {
+    type AI<'a> = RedundantCustomVerifierAI;
+
+    fn new<'a>(
+        _env: &CompilationEnv,
+        _program: &'a Program,
+        context: &'a CFGContext<'a>,
+        _init_state: &mut <Self::AI<'a> as SimpleAbsInt>::State,
+    ) -> Option<Self::AI<'a>> {
+        let Some(_) = &context.module else {
+            return None
+        };
+        let MemberName::Function(fn_name) = context.member else {
+            return None;
+        };
+
+        Some(RedundantCustomVerifierAI {
+            fn_name_loc: fn_name.loc,
+        })
+    }
+}
+
+impl SimpleAbsInt for RedundantCustomVerifierAI {
+    type State = State;
+    type ExecutionContext = ExecutionContext;
+
+    fn finish(&mut self, _final_states: BTreeMap<Label, State>, diags: Diagnostics) -> Diagnostics {
+        diags
+    }
+
+    fn start_command(&self, _: &mut State) -> ExecutionContext {
+        ExecutionContext {
+            diags: Diagnostics::new(),
+        }
+    }
+
+    fn finish_command(&self, context: ExecutionContext, _state: &mut State) -> Diagnostics {
+        let ExecutionContext { diags } = context;
+        diags
+    }
+
+    fn exp_custom(
+        &self,
+        _context: &mut ExecutionContext,
+        _state: &mut State,
+        _e: &Exp,
+    ) -> Option<Vec<Value>> {
+        None
+    }
+
+    fn call_custom(
+        &self,
+        context: &mut ExecutionContext,
+        _state: &mut State,
+        _loc: &Loc,
+        return_ty: &Type,
+        f: &ModuleCall,
+        args: Vec<Value>,
+    ) -> Option<Vec<Value>> {
+        if let Some((_, _, fname)) = PRIVATE_OBJ_FUNCTIONS
+            .iter()
+            .find(|(addr, module, fun)| f.is(addr, module, fun))
+        {
+            if let Value::LocalObjWithStore(obj_addr_loc) = args[0] {
+                let msg = format!(
+                    "Potential unnecessary implementation of a custom {} function.",
+                    fname
+                );
+                let action = if *fname == TRANSFER_FUN {
+                    "transferred"
+                } else if *fname == SHARE_FUN {
+                    "shared"
+                } else {
+                    "frozen"
+                };
+                let uid_msg = format!("Instances of a type with a store ability can be {} using \
+                                       public_{} function which often makes calling the (private) {} \
+                                       function variant in the module defining this type unnecessary", action, fname, fname);
```

Comment:
Maybe we should change the wording here. To me its less that these are unnecessary, but rather dangerous? Since you cannot guarantee that your custom transfer function will be called.

Reply:
Good point. Please take a look at the new take.